### PR TITLE
New option called `"best"` for `args.save_strategy`.

### PR DESCRIPTION
**Pull request by seanswyi**
# What does this PR do?

Addresses https://github.com/huggingface/transformers/issues/31626.

Adds a new option called `"best"` for `TrainingArguments.save_strategy` which saves the model checkpoint each time a new best performance is achieved.

### Details

1. The previous `_save_checkpoint` method was in charge of not only saving the model checkpoint but also determining the best metric and best checkpoint. The logic for determining a new best metric was separated out into the `_determine_best_metric` method.
2. `_determine_best_metric` is called after every evaluation inside of `_maybe_log_save_evaluate`. The return value `new_best_metric` is used to determine whether or not a new best metric has been achieved, and if the save strategy is `"best"` then the TrainerControl's `should_save` flag is switched on.
    * Contrary to what I initially thought, `best_metric` does not seem to be tracked by default. Rather, it's only tracked when `args.metric_for_best_model` is provided. I believe that a best metric of some sort should always be tracked, and therefore if a value is not provided then the validation loss is used to determine a new best.
3. A new object called `SaveStrategy` was created in `trainer_utils` that adds a new attribute called `BEST` to the previous `IntervalStrategy`.

I'm not sure if I like the rather "hack-y" way that I implemented this by manually switching the TrainerControl's `should_save` flag rather than delegating it to the callback handler like the other flags are dealt with. The problem is that the flags are normally updated before calling `_maybe_log_save_evaluate` inside of the inner training loop, which means there's no way for us to determine whether or not a new best metric has been achieved with the current logic. I'm not sure if I'm making sense, but I'm open to any other suggestions.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

@muellerzr @SunMarc

### CODE REVIEW

**Comment by amyeroberts** [comment]
Gentle ping @muellerzr @SunMarc

File: src/transformers/trainer.py [hunk 1]
```diff
@@ -2796,9 +2796,16 @@ def _maybe_log_save_evaluate(self, tr_loss, grad_norm, model, trial, epoch, igno
         metrics = None
         if self.control.should_evaluate:
             metrics = self._evaluate(trial, ignore_keys_for_eval)
+            new_best_metric = self._determine_best_metric(metrics=metrics, trial=trial)
+
+            if self.args.save_strategy == SaveStrategy.BEST:
+                if new_best_metric:
+                    self.control.should_save = True
+                else:
+                    self.control_should_save = False
```
Comment (muellerzr) [inline src/transformers/trainer.py:2805]: This can be refactored:

```suggestion
            self.control.should_save = new_best_metric
```

File: src/transformers/trainer.py [hunk 2]
```diff
@@ -2864,7 +2871,51 @@ def _load_rng_state(self, checkpoint):
...
+            except KeyError as exc:
+                raise KeyError(
+                    f"The `metric_for_best_model` training argument is set to '{metric_to_check}', which is not found in the evaluation metrics. "
+                    f"The available evaluation metrics are: {list(metrics.keys())}. Consider changing the `metric_for_best_model` via the TrainingArguments."
+                ) from exc
+
+            if self.args.greater_is_better:
+                operator = np.greater
+            else:
+                operator = np.less
```
Comment (muellerzr) [inline src/transformers/trainer.py:2901]: Similarly here:

```suggestion
            operator = np.greater if self.args.greater_is_better else np.less
```

**Review by muellerzr — APPROVED** [review]
While this may seem hacky to you, I think this perfectly expands our given system. Any chance you could add a few tests to this? (Over in `test_trainer.py`). Very nice work!

**Comment by seanswyi** [comment]
> While this may seem hacky to you, I think this perfectly expands our given system. Any chance you could add a few tests to this? (Over in `test_trainer.py`). Very nice work!

Thanks! And sure, I think I'll have some time to work on it over the weekend.

**Comment by seanswyi** [comment]
@muellerzr (cc. @SunMarc)

Hello. I've implemented a new test case inside of `tests/trainer/test_trainer.py` as suggested. Patches were used to control the output of the `_evaluate` method.

The first test case is when a value for `metric_for_best_model` is explicitly provided, and the second is when it's not provided and therefore defaults to the loss.

I've changed the DefaultFlowCallback object so that if `save_strategy` is either `"no"` or `"best"` then a checkpoint will not be saved at the end of training, as is the default behavior; it didn't really make sense to me that we'd want to keep only the best checkpoint but the last checkpoint is also being saved.

**Comment by seanswyi** [comment]
@muellerzr @SunMarc No rush, but just a gentle nudge/reminder! Thanks.

File: src/transformers/trainer.py [hunk 3]
```diff
@@ -2877,7 +2881,48 @@ def _load_rng_state(self, checkpoint):
...
+            run_dir = self._get_output_dir(trial=trial)
+            checkpoint_folder = f"{PREFIX_CHECKPOINT_DIR}-{self.state.global_step}"
+            output_dir = os.path.join(run_dir, checkpoint_folder)
+
+            self.state.best_metric = metric_value
+            self.state.best_model_checkpoint = output_dir
+
+            new_best_metric = True
+
+        return new_best_metric
```
Comment (SunMarc) [inline src/transformers/trainer.py:2923]: I prefer that we don't set a hardcoded metric `eval_loss` because it will be hard to find the issue. Instead, we should raise an error if the user do not set `metric_for_best_model` when using `SaveStrategy.BEST` or `load_best_model_at_end`.Hence, the last if block should only happen when inside`if self.args.metric_for_best_model is not None:`. 
LMK if this sounds good to you !

**Review by SunMarc — APPROVED** [review]
Thanks for the reminder @seanswyi ! I've added a suggestion. Please also fix the conflits and we should be good to merge !

File: src/transformers/trainer.py (hunk 3 above)
Comment (seanswyi) [reply src/transformers/trainer.py:3113]: Thanks for the suggestion, that sounds good to me. Just to make sure I understood correctly:

1. Throw an error if the user selects `SaveStrategy.BEST` or `load_best_model_at_end` but did not set a `metric_for_best_model`.
2. Remove the else block with the hard-coding of `metric_value`, hence only setting `new_best_metric = True` when `self.args.metric_for_best_model is not None`.

Let me know if there are any problems with my understanding. Otherwise, I'll go ahead and align my branch with `main` and implement the changes!

File: src/transformers/trainer.py [hunk 4]
```diff
@@ -3096,19 +3102,19 @@ def _determine_best_metric(self, metrics, trial):
...
-            run_dir = self._get_output_dir(trial=trial)
-            checkpoint_folder = f"{PREFIX_CHECKPOINT_DIR}-{self.state.global_step}"
-            output_dir = os.path.join(run_dir, checkpoint_folder)
+            if self.state.best_metric is None:
+                self.state.best_metric = float("-inf") if self.args.greater_is_better else float("inf")
+
+            if operator(metric_value, self.state.best_metric):
+                run_dir = self._get_output_dir(trial=trial)
+                checkpoint_folder = f"{PREFIX_CHECKPOINT_DIR}-{self.state.global_step}"
+                output_dir = os.path.join(run_dir, checkpoint_folder)
```
Comment (seanswyi) [inline src/transformers/trainer.py:3112]: I changed the previous code of `if self.state.best_metric is None` to explicitly setting upper and lower bounds for the values for comparisons, along with separating it out from the value comparison (i.e., `operator`).

When the `metric_for_best_model` was set to a loss value, having the `self.state.best_metric` be `None` wasn't properly comparing the values.

Let me know if there'd be a better way.

**Review by ArthurZucker — COMMENTED** [review]
Looks good, just a small nit!

File: src/transformers/trainer.py [hunk 5]
```diff
@@ -2988,9 +2994,13 @@ def _maybe_log_save_evaluate(self, tr_loss, grad_norm, model, trial, epoch, igno
         metrics = None
         if self.control.should_evaluate:
             metrics = self._evaluate(trial, ignore_keys_for_eval)
+            new_best_metric = self._determine_best_metric(metrics=metrics, trial=trial)
```
[the remark this answers is not in the document]
Comment (seanswyi) [reply src/transformers/trainer.py:2997]: Thanks for the suggestion; I also prefer this since it's more descriptive of the variable. 👍

**Review by SunMarc — APPROVED** [review]
Thanks for iterating !

File: src/transformers/trainer.py (hunk 3 above)
Comment (SunMarc) [reply src/transformers/trainer.py]: that's right. Sorry for the delay !

**Review by ArthurZucker — APPROVED** [review]
Great work, thanks for iterating!

**Comment by seanswyi** [comment]
@ArthurZucker @SunMarc @muellerzr Sorry to keep pinging you guys on this, but I'm noticing that the tests seem to be possibly stuck after the merge. Could anybody check this when they have the chance? Thanks.

https://github.com/huggingface/transformers/runs/32165247051

**Comment by SunMarc** [comment]
Indeed, thanks for noticing @seanswyi but don't worry, I check the most recent [CI](https://github.com/huggingface/transformers/actions/runs/11657144637/job/32454510816) and the test is passing. I will just cancel the runs.

**Comment by seanswyi** [comment]
> Indeed, thanks for noticing @seanswyi but don't worry, I check the most recent [CI](https://github.com/huggingface/transformers/actions/runs/11657144637/job/32454510816) and the test is passing. Can you cancel the run @ArthurZucker ? Or it will just cancel by itself after a few days ?

Ah, got it. Thanks for following up!

File: src/transformers/trainer.py [hunk 6]
```diff
@@ -419,6 +419,12 @@ def __init__(
             raise ValueError(
                 f"You have set `args.eval_strategy` to {args.eval_strategy} but you didn't pass an `eval_dataset` to `Trainer`. Either set `args.eval_strategy` to `no` or pass an `eval_dataset`. "
             )
+        if args.save_strategy == SaveStrategy.BEST or args.load_best_model_at_end:
```
Comment (shcheklein) [inline src/transformers/trainer.py:422]: @seanswyi Q: does this change mean that this logic:

https://huggingface.co/docs/transformers/main/main_classes/trainer#transformers.Seq2SeqTrainingArguments.metric_for_best_model

> ... Will default to "loss" if unspecified and load_best_model_at_end=True (to use the evaluation loss).

is broken now?

Comment (seanswyi) [reply src/transformers/trainer.py:422]: I haven't tested it yet, but it seems like it is. The docs probably need an update.

**Comment by umarbutler** [comment]
If `load_best_model_at_end = True`, the strategy cannot be set to `best`.

**Comment by umarbutler** [comment]
This solves https://github.com/huggingface/transformers/issues/35070, cheers! 🥳

File: src/transformers/trainer.py (hunk 6 above)
Comment (seanswyi) [reply src/transformers/trainer.py:422]: @shcheklein Yeah, I don't think that there's any case where `metric_for_best_model` default to loss.
